components/normalizer/src/lib.rs

```diff
@@ -1864,6 +1864,92 @@ impl DecomposingNormalizer {
 
     /// Wraps a delegate iterator into a decomposing iterator
     /// adapter by using the data already held by this normalizer.
+    ///
+    /// The [`Decomposition`] iterator will peek exactly one character
+    /// ahead of the character being decomposed, allowing the caller
+    /// to track the source character in the input string.
+    ///
+    /// # Examples
+    ///
+    /// Use a cursor to keep track of indices in the source string:
+    ///
+    /// ```
+    /// use icu_normalizer::DecomposingNormalizer;
+    /// use std::cell::RefCell;
+    /// use std::str::Chars;
+    ///
+    /// /// Struct with information on the source character being processed.
+    /// #[derive(Default)]
+    /// struct DecompositionCursorStatus {
+    ///     peeked_char: Option<char>,
+    ///     current_char: Option<char>,
+    ///     offset: usize,
+    /// }
+    ///
+    /// /// Struct implementing `Iterator<char>` with a RefCell reference
+    /// /// to the DecompositionCursorStatus
+    /// struct DecompositionCursor<'a>(
+    ///     Chars<'a>,
+    ///     &'a RefCell<DecompositionCursorStatus>,
+    /// );
+    ///
+    /// impl Iterator for DecompositionCursor<'_> {
+    ///     type Item = char;
+    ///     #[inline]
+    ///     fn next(&mut self) -> Option<char> {
+    ///         let mut cell = self.1.borrow_mut();
+    ///         if let Some(ch) = cell.current_char {
+    ///             cell.offset += ch.len_utf8();
+    ///         }
+    ///         cell.current_char = cell.peeked_char;
+    ///         cell.peeked_char = self.0.next();
+    ///         cell.peeked_char
+    ///     }
+    /// }
+    ///
+    /// // Set up the DecomposingNormalizer
+    /// let normalizer = DecomposingNormalizer::new_nfd();
+    /// let input = "Šéårçĥ réšûļţš";
+    /// let cell = RefCell::new(Default::default());
+    /// let cursor = DecompositionCursor(input.chars(), &cell);
+    /// let mut iter = normalizer.normalize_iter(cursor);
+    ///
+    /// // Test the output. get_next() is a helper function that
+    /// // fetches the next decomposed char when invoked.
+    /// let mut get_next = || {
+    ///     (
+    ///         iter.next().unwrap(),
+    ///         cell.borrow().current_char.unwrap(),
+    ///         cell.borrow().offset,
+    ///     )
+    /// };
+    ///
+    /// assert_eq!(get_next(), ('S', 'Š', 0));
+    /// assert_eq!(get_next(), ('\u{30C}', 'Š', 0));
```

Comment:
question: Is the reason why the offset only jumps at most by 2 in this example because all of the characters are in a precomposed form in the original input string in the range U+0080 <= ch < U+0800 ? If so, then optional: it might be interesting to append to the input string something in the upper half of the BMP, and maybe something beyond the BMP.

Reply:
Interesting, I can add non-BMP code points to this example.

I was also hoping maybe you could shed some light on this behavior. Is it always guaranteed that the iterator peeks one code point ahead, as stated in this PR?